app/Insightly/Resources/InsightlyOpportunityResource.php

```diff
@@ -82,4 +83,40 @@ public function linkContact(int $opportunityId, int $contactId, ContactType $con
 
         $this->insightlyClient->sendRequest($request);
     }
+
+    /**
+     * @throws ContactCannotBeUnlinked
+     */
+    public function unlinkContact(int $opportunityId, int $contactId): void
+    {
+        $getLinksRequest = new Request(
+            'GET',
+            'Opportunities/' . $opportunityId . '/Links',
+        );
+        $getLinksResponse = $this->insightlyClient->sendRequest($getLinksRequest);
+
+        $opportunityLinksAsArray = Json::decodeAssociatively($getLinksResponse->getBody()->getContents());
+
+        $linkId = null;
+        foreach ($opportunityLinksAsArray as $opportunityLink) {
+            $objectId = $opportunityLink['OBJECT_ID'];
+            $linkName = $opportunityLink['LINK_OBJECT_NAME'];
+            $linkObjectId = $opportunityLink['LINK_OBJECT_ID'];
+
+            if ($objectId === $opportunityId && $linkName === 'Contact' && $linkObjectId === $contactId) {
+                $linkId = $opportunityLink['LINK_ID'];
+            }
+        }
+
+        if ($linkId === null) {
+            throw new ContactCannotBeUnlinked();
```

Comment:
Not sure we need to throw because that will result in a failed job that can never be fixed. Maybe just logging this as a warning is enough?

Reply:
I think we need to throw here, as we want to unlink it, but it can't be done. If we don't check the linkId the code will fail. But you're right it shouldn't make the job fail. I'll add a catch statement in the listener.